tests/frame_manager_test.go

```diff
@@ -0,0 +1,87 @@
+package tests
+
+import (
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/xk6-browser/common"
+)
+
+//nolint: funlen
+func TestWaitForFrameNavigationWithinDocument(t *testing.T) {
+	t.Parallel()
+
+	navHTML := `
+<html>
+  <head>
+    <title>Navigation test within the same document</title>
+  </head>
+  <body>
+    <a id="nav-history" href="#">Navigate with History API</a>
+    <a id="nav-anchor" href="#anchor">Navigate with anchor link</a>
+    <div id="anchor">Some div...</div>
+    <script>
+      const el = document.querySelector('a#nav-history');
+      el.addEventListener('click', function(evt) {
+        evt.preventDefault();
+        history.pushState({}, 'navigated', '/nav2');
+      });
+    </script>
+  </body>
+</html>
+`
+
+	testCases := []struct {
+		name, selector string
+	}{
+		{name: "history", selector: "a#nav-history"},
+		{name: "anchor", selector: "a#nav-anchor"},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			tb := newTestBrowser(t, withHTTPServer())
+			tb.withHandler("/nav", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Set("Content-Type", "text/html")
+				_, err := w.Write([]byte(navHTML))
+				require.NoError(t, err)
+			}))
+			p := tb.NewPage(nil)
+
+			resp := p.Goto(tb.URL("/nav"), nil)
+			require.NotNil(t, resp)
+
+			el := p.Query(tc.selector)
+			require.NotNil(t, el)
+			// A click right away could possibly trigger navigation before we
+			// had a chance to call WaitForNavigation below, so give it some
+			// time to simulate the JS overhead, waiting for XHR response, etc.
+			time.AfterFunc(200*time.Millisecond, func() {
+				el.Click(nil)
+			})
+
+			done := make(chan struct{})
+			go func() {
+				defer close(done)
+				require.NotPanics(t, func() {
+					p.WaitForNavigation(tb.rt.ToValue(&common.FrameWaitForNavigationOptions{
+						Timeout: 1000, // 1s
+					}))
+				})
+			}()
+
+			select {
+			case <-done:
+			case <-time.After(2 * time.Second):
+				// WaitForNavigation is stuck?
+				close(done)
+				t.Fatal("Test timed out")
+			}
```

Comment:
:man_shrugging: I'm closing it for completion, it serves no functional purpose.

I prefer closing a channel to signal "all work is done" rather than sending an empty struct, but since there's a hypothetical chance that a double close could happen, sure, I'll change it. Though this is close to being a style suggestion ;)

> `time.Second` should be enough, I tested it with `count=50`. Also tested it for `WaitForNavigation.Timeout` (250 ms).

On your machine maybe. I prefer using relaxed timings since this will be running on underprovisioned CI machines. Both tests finish in ~0.6s on my machine, so it shouldn't have much impact for you locally.

Also, we can't use `time.Millisecond` or `time.Second` directly as an argument for `Timeout`, because it expects a millisecond integer value. We might want to change this to use duration strings instead, e.g. `200ms`, `1s`, etc.

Reply:
Resolved by 52d5a9c?